api/agent/agent.go

```diff
@@ -188,6 +188,8 @@ func (a *agent) Submit(callI Call) error {
 	ctx := call.req.Context()
 
 	span, ctx := opentracing.StartSpanFromContext(ctx, "agent_submit")
+	span.SetBaggageItem("fn_appname", callI.Model().AppName)
+	span.SetBaggageItem("fn_path", callI.Model().Path)
```

Comment:
would  be nice to have these in enclosing context, too, and for every type of request. we're spawning spans for every requests [here](https://github.com/fnproject/fn/blob/master/api/server/server.go#L103) -- `loggerWrap` in that same file has logic for extracting the `app_name` and `path` from the `*http.Request` that we could reuse, and only set if non-empty (things like `/v1/apps` won't have `path`, e.g.)

Reply:
I'd like to clarify this suggestion. Are you asking for me to set the `fn_appname` and `fn_path` baggage items in the `serve_http` span as well, by extracting the info from the request? Yes, I could do that. 

However this would need to be additional to setting them on the `agent_submit` span, since that span is not a child span of `serve_http` span and so would not "inherit" the baggage items.

Reply:
> so would not "inherit" the baggage items.

good point, for async I think this is true. for sync, it should be a child of. I'm guessing that we can set in both `serve_http` and `agent_submit` without much penalty for the duplicate addition in the sync case for the latter? and yes, that is what I meant. `serve_http` will cover every endpoint so we can see e.g. who is spamming the calls list endpoint, think it will be convenient.

Reply:
Right. The main issue is that some calls to `serve_http` have a appname and path, and some do not. Prometheus requires a given metric to have the same labels every time (otherwise the go client panics with "inconsistent label cardinality"). So this would force us to change the prometheus collector to set the labels `fn_appname` and `fn_path` on *every* metric, using an empty string when a value was not available from the span. That makes the collector less generic.

Here's what `traceWrap` would look like (for the `server_http` span). It's neat enough...
```
func traceWrap(c *gin.Context) {
	// try to grab a span from the request if made from another service, ignore err if not
	wireContext, _ := opentracing.GlobalTracer().Extract(
		opentracing.HTTPHeaders,
		opentracing.HTTPHeadersCarrier(c.Request.Header))

	// Create the span referring to the RPC client if available.
	// If wireContext == nil, a root span will be created.
	// TODO we should add more tags?
	serverSpan := opentracing.StartSpan("serve_http", ext.RPCServerOption(wireContext), opentracing.Tag{Key: "path", Value: c.Request.URL.Path})
	if appName := c.Param(api.CApp); appName != "" {
		serverSpan.SetBaggageItem("fn_appname", appName)
	}
	if routePath := c.Param(api.CRoute); routePath != "" {
		serverSpan.SetBaggageItem("fn_path", routePath)
	}
	defer serverSpan.Finish()

	ctx := opentracing.ContextWithSpan(c.Request.Context(), serverSpan)
	c.Request = c.Request.WithContext(ctx)
	c.Next()
}
```

And here's what the collector might look like, with hardcoded label names:
```
// Return the HistogramVec corresponding to the specified spanName.
// If a HistogramVec does not already exist for specified spanName then one is created and configured with the specified labels
// otherwise the labels parameter is ignored.
func (pc PrometheusCollector) getHistogramVecForSpanName(spanName string, labels []string) *prometheus.HistogramVec {
	thisHistogramVec, found := pc.histogramVecMap[spanName]
	if !found {
		thisHistogramVec = prometheus.NewHistogramVec(
			prometheus.HistogramOpts{
				Name: "fn_span_" + spanName + "_duration_seconds",
				Help: "Span " + spanName + " duration, by span name",
			},
			labels,
		)
		pc.histogramVecMap[spanName] = thisHistogramVec
		prometheus.MustRegister(thisHistogramVec)
	}
	return thisHistogramVec
}

// PrometheusCollector implements Collector.
func (pc PrometheusCollector) Collect(span *zipkincore.Span) error {

	// all metrics have the same label names
	labelValueMap := &map[string]string{
		"fn_appname": "",
		"fn_path":    "",
	}

	labelKeys := []string{"fn_appname", "fn_path"}

	// fill labelValueMap with any label values found in the span
	fillWithLabelValues(span, labelValueMap)

	pc.getHistogramVecForSpanName(span.GetName(), labelKeys).With(*labelValueMap).Observe((time.Duration(span.GetDuration()) * time.Microsecond).Seconds())
	return nil
}

func fillWithLabelValues(span *zipkincore.Span, labelMap *map[string]string) {

	var keys []string

	// extract any tags whose key matches one of the keys in labelMap, and update the corresponding value
	binaryAnnotations := span.GetBinaryAnnotations()
	for _, thisBinaryAnnotation := range binaryAnnotations {
		key := thisBinaryAnnotation.GetKey()
		if thisBinaryAnnotation.GetAnnotationType() == zipkincore.AnnotationType_STRING {
			if _, ok := (*labelMap)[key]; ok {
				keys = append(keys, key)
				value := string(thisBinaryAnnotation.GetValue()[:])
				(*labelMap)[key] = value
			}
		}
	}
}

Any thoughts before I commit and push these changes?
```

Reply:
Seems to look OK in Grafana (currently the legend says fn_path(fn_appname) for the series where these labels are empty: I expect this could be changed to show something better)
![grafana - fn tracing span data - mozilla firefox 24102017 184647](https://user-images.githubusercontent.com/6053562/31959151-d4c4ae52-b8eb-11e7-80d6-d1f7fbbda735.jpg)

Reply:
> this would force us to change the prometheus collector to set the labels fn_appname and fn_path on every metric, using an empty string when a value was not available from the span.

while this is pretty unfortunate, this seems like a sensible solution to me. For child spans of `serve_http` they will be consistent, generally, I would assume (maybe some db spans this isn't true). I'm assuming we will be able to pretty easily ignore empty strings like this from whatever tool we're using to read the metrics, anyway, this seems fine to me. 

The alternative I guess is to add new spans to each of our HTTP handlers and there we can pick and choose whether the endpoint will have `path` and `app_name` available from the span and it will then be consistent, though it's somewhat concerning I guess that we couldn't have e.g. span `get_calls` with label `app_name` and no `path` and also `get_calls` with both `app_name` and `path`, as both of these are perfectly valid and use the same db methods and endpoints. so maybe we will need to inject empty strings anyway and what you have proposed is preferable?

Reply:
I've realised that we can avoid the extra complexity of the collector by changing `traceWrap` to set `fn_appname` and `fn_path` to empty strings when there are no values to set them to. Then I think we can leave the collector as it is now (phew). 

```
	serverSpan := opentracing.StartSpan("serve_http", ext.RPCServerOption(wireContext), opentracing.Tag{Key: "path", Value: c.Request.URL.Path})
	if appName := c.Param(api.CApp); appName != "" {
		serverSpan.SetBaggageItem("fn_appname", appName)
	} else {
		serverSpan.SetBaggageItem("fn_appname", "")
	}
	if routePath := c.Param(api.CRoute); routePath != "" {
		serverSpan.SetBaggageItem("fn_path", routePath)
	} else {
		serverSpan.SetBaggageItem("fn_path", "")
	}
	defer serverSpan.Finish()
```

Reply:
+1 -- I guess `serverSpan.SetBaggageItem("fn_appname", c.Param(api.CApp))` would probably do then fwiw since we are setting in both cases

Reply:
Yes, indeed.